File: src/day7.py


class Bag:
    all_ = dict()
    stuff = set()

    def __init__(self, name):
        self.name = name
        self.upper = []
        self.contains = []

    def add_upper(self, other):
        if other not in self.all_:
            self.all_[other] = Bag(other)

        self.upper.append(self.all_[other])

    def add_contain(self, other, count):
        self.contains.append((self.all_[other], count))

    def get_upper(self):
        print(self.name)
        self.stuff.add(self.name)
        if not self.upper:
            return {self.name}

        temp = {self.name}
        for bag in self.upper:
            temp.update(bag.get_upper())

        return temp

    def contents(self):
        if not self.contains:
            return 0

        total_count = 0
        for bag, count in self.contains:
            total_count += count * (bag.contents() + 1)

        return total_count


def first():

    for line in open("data/day7.txt"):
        bag, rest = line.split(" contain ")
        bag = bag[:-1]

        rest = rest.strip()

        if bag not in Bag.all_:
            Bag.all_[bag] = Bag(bag)

        if rest == "no other bags.":
            continue

        for item in rest.split(", "):
            bag_ = item[2:].strip(".").rstrip("s")
            if bag_ not in Bag.all_:
                Bag.all_[bag_] = Bag(bag_)

            Bag.all_[bag_].add_upper(bag)

            Bag.all_[bag].add_contain(bag_, int(item[0]))

    print(len(Bag.all_["shiny gold bag"].get_upper()), len(Bag.stuff))
    print(Bag.all_["shiny gold bag"].contents())


if __name__ == '__main__':
    first()

File: src/day7.rs
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{BufReader, BufRead};


// Garbage language. Please tell me how to do this

#[derive(Hash, Eq, PartialEq, Debug)]
struct Bag<'a> {
    can_contain_this: Vec<&'a Bag<'a>>,
    name: String,
}


impl Bag<'_> {
    fn new(name: String) -> Bag<'static> {
        return Bag{
            can_contain_this: Vec::new(),
            name
        };
    }

    fn add_upper(&mut self, other: Bag<'_>){
        self.can_contain_this.push(&other);
    }
}

fn first(){
    let file = match File::open("data/day7.txt") {
        Ok(s) => s,
        Err(_) => panic!("File doesn't exist")
    };
    let reader = BufReader::new(file);

    let mut bags: HashMap<String, Bag> = HashMap::new();

    for line in reader.lines() {
        let string = line.unwrap();
        let asd: Vec<&str> = string.split(" contain ").collect();

        if asd[1] == "no other bags."{
            continue;
        }

        let first = asd[0][0..&asd[0].len() - 1].to_string();

        if !bags.contains_key(&first) {
            bags.insert(first.to_string(), Bag::new(first));
        }


        for bag in asd[1].split(", ") {
            let a = bag[2..].to_string();
            if !bags.contains_key(&a) {
                bags.insert(a.to_string(), Bag::new(a));
            }
            let mut temp = bags.get_mut(&a).unwrap();
            let t = bags.get(&first).unwrap();
            temp.add_upper(t);
        }
    }

    let mut seen = HashSet::new();

    for i in &bags["shiny gold bag"].can_contain_this {
        println!("{}", i.name);
    }
}


fn main() {
    first();
}